File: frontend/src/api/streak.ts
// src/api/streak.ts
//
// Daily streak gamification (Beli/Duolingo-style). See the backend's
// streak_service.py for the actual day-boundary logic -- this just sends
// the device's current IANA timezone name so the server can compute the
// correct local calendar day; the instant itself always comes from the
// server, never the device clock.
import { client } from './client';

export interface Streak {
  current_streak: number;
  longest_streak: number;
  last_active_date: string | null;
}

function deviceTimezone(): string | undefined {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone;
  } catch {
    return undefined;
  }
}

export async function fetchMyStreak(): Promise<Streak> {
  const { data } = await client.get<Streak>('/api/v1/streak/me');
  return data;
}

// Call this on app open/foreground -- idempotent for the same calendar
// day, so it's safe to call more than once.
export async function pingStreak(): Promise<Streak> {
  const { data } = await client.post<Streak>('/api/v1/streak/ping', {
    timezone: deviceTimezone(),
  });
  return data;
}

File: backend/app/api/v1/routes/streak.py
# FILE: backend/app/api/v1/routes/streak.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Body, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.auth import require_api_key
from app.core.rate_limit import rate_limit
from app.core.user_auth import get_current_user_id
from app.db.session import get_db
from app.services.social.streak_service import get_streak, record_activity

router = APIRouter(prefix="/streak", tags=["streak"])


class StreakOut(BaseModel):
    current_streak: int
    longest_streak: int
    last_active_date: Optional[str] = None


class StreakPingRequest(BaseModel):
    # IANA timezone name, e.g. "America/Los_Angeles" -- used only to
    # compute the correct calendar day for the streak boundary. See
    # streak_service's module docstring for why the instant itself still
    # only ever comes from the server, never this value.
    timezone: Optional[str] = None


def _to_out(state) -> StreakOut:
    return StreakOut(
        current_streak=state.current_streak,
        longest_streak=state.longest_streak,
        last_active_date=state.last_active_date.isoformat() if state.last_active_date else None,
    )


@router.get(
    "/me",
    response_model=StreakOut,
    dependencies=[Depends(rate_limit), Depends(require_api_key)],
)
def get_my_streak(
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
) -> StreakOut:
    """Read-only -- does not record activity. Call POST /streak/ping for that."""
    return _to_out(get_streak(db, user_id=user_id))


@router.post(
    "/ping",
    response_model=StreakOut,
    dependencies=[Depends(rate_limit), Depends(require_api_key)],
)
def ping_streak(
    body: StreakPingRequest = Body(default=StreakPingRequest()),
    db: Session = Depends(get_db),
    user_id: str = Depends(get_current_user_id),
) -> StreakOut:
    """
    Called by the app on open/foreground. Idempotent for the same
    calendar day -- safe to call more than once per day.
    """
    return _to_out(record_activity(db, user_id=user_id, client_timezone=body.timezone))
